### PR TITLE
feat: added synchronization from CMS to repo

### PR DESCRIPTION
**Pull request by nickytonline**
## Description

Synchronizes data coming from our CMS via a Webhook to the plugins repo that dispatches a GitHub action. 

Steps the action performs:

- Creates and switches to a new branch with the following naming convention, "`sync_cms_to_plugins_{date in milliseconds}"`
- Updates site/plugins.json and commits it to the branch
- Creates a pull request with the title `"chore: cms to repo sync"` and two labels: `cms_sync` and `automerge`
- Assuming the PR passes all checks, it merges automatically

## Related Issues and Documents

- Relates to #820 
- Closes https://github.com/netlify/pod-ecosystem-frameworks/issues/223

**Are you adding a plugin or updating one?**

- [ ] Adding a plugin
- [ ] Updating a plugin
- [x] Automation

**Have you completed the following?**

- [x] Read and followed the [plugin author guidelines](https://github.com/netlify/plugins/blob/main/docs/guidelines.md).
- [ ] Included all [required fields](https://github.com/netlify/plugins/blob/main/docs/CONTRIBUTING.md#required-fields) in your entry.
- [ ] Tested the plugin [locally](https://docs.netlify.com/cli/get-started/#run-builds-locally) and [on Netlify](https://docs.netlify.com/configure-builds/build-plugins/#install-a-plugin), using the plugin version stated in your entry.

**Test plan**
Please add a link to a successful public deploy log using the stated version of the plugin. Include any other context reviewers might need for testing.

- The Webhook sends the data now, as can be seen in this instance of the workflow that is currently on main, https://github.com/netlify/plugins/actions/runs/3146754307/jobs/5115554802. The message `Synchronizing changes to plugins repo...` in the logs means there was data sent if you look at the associated code on main.

```
// This will be replaced in a follow up PR  with the actual sync logic
if (changes && Object.keys(changes).length !== 0) {
  console.log('Synchronizing changes to plugins repo...')
}
```

https://github.com/netlify/plugins/blob/main/bin/sync_cms_to_plugins_repo.js#L6-L9

- With some test changed data, I used that as the `process.env.CMS_CHANGES` to test locally and it creates the PR with the correctly named branch and tags. I've only been able to test this on my own repository since this is not merged to main yet.

Two sinks dripping

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -46,6 +46,7 @@
     "normalize-node-version": "^11.0.0",
     "pacote": "^13.0.0",
     "semver": "^7.0.0",
+    "tsx": "^3.9.0",
```
Comment (nickytonline) [inline package.json:49]: This has already been added in #820 but until that's merged, I need it in this PR.

File: bin/sync_cms_to_plugins_repo.js [hunk 2]
```diff
@@ -1,11 +1,24 @@
...
-const changes = JSON.parse(process.env.CMS_CHANGES)
+import plugins from '../site/plugins.json'
 
-console.log('Checking for CMS updates...')
+import { updatePlugins } from './utils.js'
 
-// This will be replaced in a follow up PR  with the actual sync logic
-if (changes && Object.keys(changes).length !== 0) {
-  console.log('Synchronizing changes to plugins repo...')
+// TODO:eslint-disable-next-line n/prefer-global/process
```
Comment (nickytonline) [inline bin/sync_cms_to_plugins_repo.js:5]: This is commented out temporarily so I can test the GH action with fake changes.

File: package.json (hunk 1 above)
Comment (nickytonline) [inline package.json:49]: This has already been added in https://github.com/netlify/plugins/pull/820 but until that's merged, I need it in this PR.

Comment (nickytonline) [inline package.json:49]: This has already been added in https://github.com/netlify/plugins/pull/820 but until that's merged, I need it in this PR.

File: bin/sync_cms_to_repo.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+echo "Creating PR \"$PR_TITLE\" for branch $BRANCH_NAME"
+
+// This is the only file we want to commit
+git add site/plugins.json
+
+// See if we have any changes. We should.
+if [[ -n "$(git status --porcelain)" ]]; then
+  git commit -m $PR_TITLE
+  git push origin $BRANCH_NAME
+  gh pr create --title $PR_TITLE --body "This is an automated PR to sync the CMS to the repo" --label "cms_sync" --label "automerge"
```
Comment (nickytonline) [inline bin/sync_cms_to_repo.sh:19]: I created a label in the repo called `cms_sync` for filtering of automated PRs, and I also add the `automerge` label which should auto merge the PR that gets generated.

File: types/plugins.d.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
+export interface SanityBuildPluginEntity {
```
Comment (nickytonline) [inline types/plugins.d.ts:1]: This file is in #820, but there is also a small update in here. Once #820 is merged, the changes here should be a one liner.

File: bin/utils.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    status,
+    compatibility,
+  }
+}
+
+const stripNullifiedFields = (plugin) => {
+  // If plugin status is undefined, it means the plugin is active, so the field is omitted
+  if (!plugin.status) {
+    // eslint-disable-next-line no-param-reassign
+    delete plugin.status
```
Comment (estephinson) [inline bin/utils.js:35]: First PR i've ever seen with `delete` used, I'm proud

File: bin/sync_cms_to_repo.sh (hunk 3 above)
Comment (estephinson) [reply bin/sync_cms_to_repo.sh:19]: noice

**Review by estephinson — APPROVED** [review]
amazing stuff, love it

File: bin/utils.js (hunk 5 above)
Comment (nickytonline) [reply bin/utils.js:35]: I could have maybe done this another way, but we don't care about immutability in this scenario.

File: bin/sync_cms_to_repo.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+// This is the only file we want to commit
+git add site/plugins.json
+
+// See if we have any changes. We should.
+if [[ -n "$(git status --porcelain)" ]]; then
+  git commit -m $PR_TITLE
+  git push origin $BRANCH_NAME
+  # TODO: Uncomment the line below and delete the one below it once we're happy that this is working well in the test environment
+  # gh pr create --title $PR_TITLE --body "This is an automated PR to sync the CMS to the repo" --label "cms_sync" --label "automerge"
+  gh pr create --title $PR_TITLE --body "DO NOT MERGE, TESTING SYNC: This is an automated PR to sync the CMS to the repo" --label "cms_sync"
```
Comment (nickytonline) [inline bin/sync_cms_to_repo.sh:21]: @estephinson @tzmanics, I'll put this back to this once we're happy with the sync in the generated PRs that this action does while testing with the test data.

```suggestion
  gh pr create --title $PR_TITLE --body "This is an automated PR to sync the CMS to the repo" --label "cms_sync" --label "automerge"
```

**Review by tzmanics — APPROVED** [review]
well done

**Pull request merged**